SQL Server Assessment

noyasarko opened the pull request:
feat: Add SQL Server assessment functionality

The code changes include adding SQL Server assessment functionality to the script. This enhancement allows for assessing SQL Servers in Microsoft Azure subscriptions.

Closes #31 #32

Commit:
```text
feat: Add SQL Server assessment functionality

The code changes include adding SQL Server assessment functionality to the script. This enhancement allows for assessing SQL Servers in Microsoft Azure subscriptions.

Nicwalle commented on SQL/sqlRules.json:
Great set of tests. To make it actionable, can you please add some links with explanation on how/why one should do it. The same way as it is done in the [AKS one](https://github.com/beluxappdev/health-check/blob/main/AKS/aksRules.json).

noyasarko replied:
On it, tho this might take a while for me, prioritizing handling other parts of the code first

Nicwalle commented on SQL/sqlRules.json:
```diff
@@ -0,0 +1,47 @@
+    "expected": true,
+    "explanation": "The server should have firewall rules enabled, no more than 10.",
+    "function": "hasFirewallRulesEnabled"
+    },
```
Looking at the explanation, that looks like 2 different checks that could be made. 1 for "has FW rules enabled" and 1 for "no more than 10 rules". Agreed?

noyasarko replied:
so the way this works is that the enablement pertains to the "selected networks" being chosen under Networking. The 10 is for the firewall rules, so I will indeed split this into 2.

Nicwalle commented on SQL/SQLServerCheck.psm1:
```diff
@@ -0,0 +1,136 @@
+            }
+        }
+        return $true
+    }
```
I had a discussion with Frederik on this one, there can apparently be multiple rules which each contain a range. 
The way you could tackle that `firewallHasMax10AuthorizedIPs` is by counting them. E.g. (to be tested)

```suggestion
    [bool] firewallHasMax10AuthorizedIPs() {
        $totalIPCount = 0
        foreach ($rule in $this.ServerObject.firewallRules) {
            $startIP = $rule.startIpAddress.Split(".")[3]
            $endIP = $rule.endIpAddress.Split(".")[3]
            $totalIPCount = $totalIPCount + [math]::Abs([int]$endIP - [int]$startIP) + 1
        }
        return $totalIPCount -le 10
    }
```
Notice the +1 as, for example 10.0.0.3 -> 10.0.0.5 contains 3 IP (3, 4, 5) but 5-3=2. Hence the need for a +1 I guess.

Nicwalle commented on StartAKSClusterAssessment.ps1:
```diff
@@ -65,17 +92,46 @@ foreach ($currentSubscription in $subscriptions) {
+    # Convert back to JSON if needed
+    $jsonSQLservers = $servers | ConvertTo-Json -Depth 10
+    $fullservers = $jsonSQLservers | ConvertFrom-Json 
+}
```
This logic should be part of the firewall IP checks themselves (with some potential re-usable function which can cache the results if it is called often) and not alter the object. As part of a check, you can run az cli command in a check.

noyasarko commented on StartAKSClusterAssessment.ps1:
```diff
@@ -55,7 +57,32 @@ Write-Host "Output Path: $OutPath" -ForegroundColor Blue
 
 Start-Transcript -Path "$OutPath\log_$today.txt"
 
-Write-Host "******** Welcome to the Microsoft Azure Kubernetes Cluster assessment" -ForegroundColor Cyan
```
I literally just added stars at the end... thats the only difference

noyasarko commented on StartAKSClusterAssessment.ps1:
```diff
@@ -65,17 +92,46 @@ foreach ($currentSubscription in $subscriptions) {
     Write-Host "***** Assessing the subscription $($currentSubscription.displayName) ($($currentSubscription.id)..." -ForegroundColor Cyan
     az account set -s $currentSubscription.SubscriptionId --only-show-errors 
 
-    $jsonAksClusters = az aks list -o json --only-show-errors 
```
I honestly don't know what this piece of code is doing here, I believe the entire section of this red zone 68-75 should be removed

noyasarko commented on StartAKSClusterAssessment.ps1:
```diff
@@ -65,17 +92,46 @@ foreach ($currentSubscription in $subscriptions) {
+$servers = az sql server list -o json --only-show-errors | ConvertFrom-Json
+$firewallRules = az sql server firewall-rule list -o json --only-show-errors | ConvertFrom-Json
 
-        $aksCluster.assess().GetAllResults() | Export-Csv -Path "$OutPath\assess_$today.csv" -NoTypeInformation -Append -Delimiter $csvDelimiter
```
again AKS line in the middle of SQL assessment

noyasarko commented on StartAKSClusterAssessment.ps1:
```diff
@@ -65,17 +92,46 @@ foreach ($currentSubscription in $subscriptions) {
+$firewallRules = az sql server firewall-rule list -o json --only-show-errors | ConvertFrom-Json
 
-        $aksCluster.assess().GetAllResults() | Export-Csv -Path "$OutPath\assess_$today.csv" -NoTypeInformation -Append -Delimiter $csvDelimiter
+# Assuming firewall rules contain a property that can be used to match them to their respective servers,
```
lil background, so I needed info from a different code so I combined it in to 1 json and use that as source. Additionally, SQL DB section is included but as comment bcz I'm not sure yet

Nicwalle committed:
Fixed indentation

Commit:
feat: Update SQL Server assessment functionality

The code changes include updating the SQL Server assessment functionality in the script. This update improves the assessment process for SQL Servers in Microsoft Azure subscriptions.

Nicwalle commented on SQL/SQLServerCheck.psm1:
```diff
@@ -0,0 +1,130 @@
+
+    #Check if Firewall start IP and end IP difference is less than 10
+    [bool] firewallHasMax10AuthorizedIPs() {
+        $firewallRules = az sql server firewall-rule list -g  $this.ServerObject.resourceGroup -s $this.ServerObject.name -o json | ConvertFrom-Json  
```
Using the getter is recommended for future-proofness and readability
```suggestion
        $firewallRules = az sql server firewall-rule list -g  $this.getServerResourceGroup() -s $this.getServerName() -o json | ConvertFrom-Json  
```

Nicwalle commented on SQL/SQLServerCheck.psm1:
```diff
@@ -0,0 +1,130 @@
+
+    #Check if Firewall rules are enabled
+    [bool] hasMax10FirewallRulesEnabled() {
+        $firewallRules = az sql server firewall-rule list -g  $this.ServerObject.resourceGroup -s $this.ServerObject.name -o json | ConvertFrom-Json
```
Using the getter is recommended for future-proofness and readability
```suggestion
        $firewallRules = az sql server firewall-rule list -g  $this.getServerResourceGroup() -s $this.getServerName() -o json | ConvertFrom-Json
```

Nicwalle commented on SQL/SQLServerCheck.psm1:
```diff
@@ -0,0 +1,130 @@
+
+    # Checks if Auditing on server level is enabled
+    [bool] hasAuditingEnabled() {
+        $ServerAuditing =  az sql server audit-policy show --name $this.ServerObject.name --resource-group $this.ServerObject.resourceGroup -o json | ConvertFrom-Json
```
Using the getter is recommended for future-proofness and readability
```suggestion
        $ServerAuditing =  az sql server audit-policy show --name $this.getServerName() --resource-group $this.getServerResourceGroup() -o json | ConvertFrom-Json
```

Nicwalle commented on StartAKSClusterAssessment.ps1:
```diff
@@ -65,17 +92,26 @@ foreach ($currentSubscription in $subscriptions) {
-    $jsonAksClusters | Out-File -FilePath "$OutPath\raw_$today.json" -Append
-    $aksClusters = $jsonAksClusters | ConvertFrom-Json -AsHashTable
+    $jsonSQLservers = az sql server list -o json --only-show-errors 
+    $jsonSQLservers | Out-File -FilePath "$OutPath\raw_$today.json" -Append
```
```suggestion
    $jsonSQLservers | Out-File -FilePath "$OutPath\sql_server_raw_$today.json" -Append
```

Nicwalle commented on SQL/sqlRules.json:
```diff
@@ -0,0 +1,42 @@
+{
+  "SQL_Server_Auth_EntraIDIntegrated": {
+    "expected": true,
+    "explanation": "The server should be integrated with Azure Active Directory for authentication and authorization. For more information: https://learn.microsoft.com/en-us/sql/relational-databases/security/authentication-access/azure-ad-authentication-sql-server-overview?view=sql-server-ver16",
```
This link does not seem to be the right one as it is related to sql server (on-prem) and not Azure SQL. This one seems better: https://learn.microsoft.com/en-us/azure/azure-sql/database/authentication-aad-configure?view=azuresql&tabs=azure-powershell

Also, please change the reference to Azure Active Directory to Microsoft Entra ID instead as it is its new name

Nicwalle commented on SQL/sqlRules.json:
```diff
@@ -0,0 +1,42 @@
+  },
+  "SQL_Server_Auditing" : {
+    "expected": true,
+    "explanation": "The server should have auditing enabled. For more information: https://learn.microsoft.com/en-us/sql/relational-databases/security/auditing/sql-server-audit-database-engine?view=sql-server-ver16",
```
This link seems incorrect as related to SQL Server 2022. Here is one for enabling auditing for Azure SQL: https://learn.microsoft.com/en-us/azure/azure-sql/database/auditing-setup?view=azuresql

Nicwalle commented on SQL/sqlRules.json:
```diff
@@ -0,0 +1,42 @@
+  },
+  "SQL_Server_TLS_1.2": {
+    "expected": true,
+    "explanation": "The server should have TLS 1.2 enabled. For more information: https://learn.microsoft.com/en-us/sql/database-engine/configure-windows/certificate-overview?view=sql-server-ver16",
```
Link related to SQL server on-prem. Might need to change it to https://learn.microsoft.com/en-us/azure/azure-sql/database/connectivity-settings?view=azuresql&tabs=azure-portal#minimal-tls-version

Commit:
Update SQL Server assessment functionality

Nicwalle commented on SQL/sqlRules.json:
```diff
@@ -0,0 +1,42 @@
+  },
+  "SQL_Server_Firewall_Rules": {
+    "expected": true,
+    "explanation": "The server should have no more than 10 firewall rules enabled. For more information: https://learn.microsoft.com/en-us/sql/sql-server/install/configure-the-windows-firewall-to-allow-sql-server-access?view=sql-server-ver16",
```
Change for this link: https://learn.microsoft.com/en-us/azure/azure-sql/database/firewall-configure?view=azuresql

Nicwalle commented on SQL/sqlRules.json:
```diff
@@ -0,0 +1,42 @@
+  },
+    "SQL_Firewall_IP_Range": {
+    "expected": true,
+    "explanation": "The server should have a firewall IP range set, no more than 10 IPs. For more information: https://learn.microsoft.com/en-us/sql/sql-server/install/configure-the-windows-firewall-to-allow-sql-server-access?view=sql-server-ver16",
```
Change for this link: https://learn.microsoft.com/en-us/azure/azure-sql/database/firewall-configure?view=azuresql

Commit:
feat: Update SQL Server assessment functionality part 2

Commit:
Fixed spacing, as well as auditing check

Nicwalle commented on SQL/sqlRules.json:
```diff
@@ -0,0 +1,42 @@
+  },
+  "SQL_Server_Security_OutboundNetworkAccess": {
+    "expected": true,
+    "explanation": "The server should have restrictions on outbound network access disabled. For more information: https://learn.microsoft.com/en-us/sql/sql-server/install/configure-the-windows-firewall-to-allow-sql-server-access?view=sql-server-ver16 ",
```
Should be changed to https://learn.microsoft.com/en-us/azure/azure-sql/database/outbound-firewall-rule-overview

Nicwalle reviewed (changes requested):
Also check, there are still some references to Azure Active Directory that should be changed into Entra ID

Nicwalle committed:
Added StartSQLAssessment file in SQL folder

Nicwalle committed:
Deleted old StartAKSAssessment file

Nicwalle committed:
Added support for SQL in main files

Commit:
link changed

Nicwalle commented on SQL/SQLServerCheck.psm1:
```diff
@@ -0,0 +1,132 @@
+
+    #Checks if Minimum TLS version is set to 1.2
+    [bool] hasMinimumTLSVersion12() {
+        return $this.ServerObject.minimalTlsVersion -ge "1.2"
```
Check the comparison between number instead of strings

Commit:
Small edits

Commit:
change in auditing check

Commit:
sql rules changes in the naming

Nicwalle reviewed (approved):
Looks good to me. Just need to double check for the outbound check. Accepting it for now. Might need to be updated later. ENABLED/DISABLED??
